azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java

```diff
@@ -54,7 +54,8 @@
   @Override
   public void uploadExecutableFlow(ExecutableFlow flow)
       throws ExecutorManagerException {
-    flows.put(flow.getExecutionId(), flow);
+    ExecutableFlow exFlow = ExecutableFlow.createExecutableFlowFromObject(flow.toObject());
```

Comment:
> I am still having trouble understanding this problem and fix.

I guess you checked the [commit message](https://github.com/azkaban/azkaban/pull/1105/commits/ef2c53d82f9c432425a12925395843b75146a340)?

To add a bit of background to that, this test uses MockExecutorLoader instead of JDBCExecutorLoader. The uploadExecutableFlow() method saves current flow state into DB (I'm not 100% sure if there's any another reason, but at least via the DB it becomes accessible to the Azkaban web server). The mock version doesn't use a DB but local maps to keep the flows and jobs. I guess this is so that in test you could verify in the test "what would it save in DB".

Now on to the problem posed by this. The MockExecutorLoader created clones of the jobs, so (non-synchronized) job state updates didn't mess with the job nodes. But it didn't clone the flow object. And the original flow object has references to the original job nodes, too! When uploadExecutableFlow() was called, it also updated the job statuses on the _original_ ExecutableNodes based on the status values know to the FlowRunner thread at that time (see the stack trace in fix commit message to see why this causes job statuses being set). It's quite clear that to "mock the DB" this test shouldn't modify the original flows/ExecutableNodes.